components/support/nimbus-fml/src/intermediate_representation.rs

```diff
@@ -54,6 +54,140 @@ pub struct FeatureManifest {
+                }
+                self.validate_type_ref(value_type, enum_names, obj_names)
+            }
+            TypeRef::List(list_type) => self.validate_type_ref(list_type, enum_names, obj_names),
+            TypeRef::StringMap(value_type) => {
+                self.validate_type_ref(value_type, enum_names, obj_names)
+            }
+            TypeRef::Option(option_type) => {
+                if let TypeRef::Option(_) = option_type.as_ref() {
+                    return Err(FMLError::ValidationError(
+                        "Found nested optional types".into(),
+                    ));
+                }
+                Ok(())
+            }
+            _ => Ok(()),
+        }
+    }
+
+    fn validate_enum_defs(&self, enum_names: &mut HashSet<String>) -> Result<()> {
+        for enum_def in &self.enum_defs {
+            if !enum_names.insert(enum_def.name.clone()) {
+                return Err(FMLError::ValidationError(format!(
+                    "EnumDef names must be unique. Found two EnumDefs with the same name: {}",
+                    enum_def.name
+                )));
+            }
+        }
+        Ok(())
+    }
+
+    fn validate_obj_defs(&self, obj_names: &mut HashSet<String>) -> Result<()> {
+        for obj_def in &self.obj_defs {
+            if !obj_names.insert(obj_def.name.clone()) {
+                return Err(FMLError::ValidationError(format!(
+                    "ObjectDef names must be unique. Found two ObjectDefs with the same name: {}",
+                    obj_def.name
+                )));
+            }
+        }
+        Ok(())
+    }
+
+    fn validate_feature_defs(&self, feature_names: &mut HashSet<String>) -> Result<()> {
+        for feature_def in &self.feature_defs {
+            if !feature_names.insert(feature_def.name.clone()) {
+                return Err(FMLError::ValidationError(format!(
+                    "FeatureDef names must be unique. Found two FeatureDefs with the same name: {}",
+                    feature_def.name
+                )));
+            }
+            // while checking the feature, we also check that each prop is unique within a feature
+            let mut prop_names = HashSet::new();
+            for prop in &feature_def.props {
+                if !prop_names.insert(prop.name.clone()) {
+                    return Err(FMLError::ValidationError(format!(
+                        "PropDef names must be unique. Found two PropDefs with the same name: {} in the same feature_def: {}",
+                        prop.name, feature_def.name
+                    )));
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn validate_defaults(&self) -> Result<()> {
+        // TODO: Validate that the defaults of each feature
+        // in the feature manifest agree with the props
+        // There must be a default for each property
+        // either that's defined in the property itself
+        // or in the feature definition
+        Ok(())
```

Comment:
Ok, so there are two levels of defaulting emerging. I've broken these out into two separate bugs so you can land this more quickly:

* the [property level defaults][1].
* the [feature level defaults][2].

[1]: https://mozilla-hub.atlassian.net/browse/SDK-441
[2]: https://mozilla-hub.atlassian.net/browse/SDK-424